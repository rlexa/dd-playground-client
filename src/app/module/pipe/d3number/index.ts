export * from './d3number.pipe';
