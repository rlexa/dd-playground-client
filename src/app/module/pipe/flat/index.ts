export * from './flat-pipe.module';
