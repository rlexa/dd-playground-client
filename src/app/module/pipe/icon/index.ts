export * from './icon-pipe.module';
