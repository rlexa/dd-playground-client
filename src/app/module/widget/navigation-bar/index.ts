export * from './navigation-bar-items-from-route.directive';
export * from './navigation-bar.component';
