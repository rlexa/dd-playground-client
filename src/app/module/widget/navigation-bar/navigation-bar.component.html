<mat-toolbar color="primary">
  <span class="mat-toolbar-filler"></span>
  <ng-container *ngIf="items as iis">
    <nav class="mat-nav-list">
      <a
        *ngFor="let ii of iis; trackBy: trackByIndex"
        mat-button
        [style.min-width]="'auto'"
        [routerLink]="ii.route"
        routerLinkActive="active"
        [matTooltip]="ii.label"
      >
        <i class="material-icons transition-transform active-scale">{{ ii.icon | icon }}</i>
      </a>
    </nav>
    <span class="mat-toolbar-filler"></span>
  </ng-container>
</mat-toolbar>
