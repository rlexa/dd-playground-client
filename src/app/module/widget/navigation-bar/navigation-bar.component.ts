import {CommonModule} from '@angular/common';
import {ChangeDetectionStrategy, Component, Input} from '@angular/core';
import {MatButtonModule} from '@angular/material/button';
import {MatToolbarModule} from '@angular/material/toolbar';
import {MatTooltipModule} from '@angular/material/tooltip';
import {RouterModule} from '@angular/router';
import {IconPipe} from '../../pipe/icon';

export interface NavigationBarItem {
  icon: string;
  label: string;
  route: string;
}

@Component({
  selector: 'app-navigation-bar',
  templateUrl: './navigation-bar.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
  standalone: true,
  imports: [CommonModule, RouterModule, MatButtonModule, MatToolbarModule, MatTooltipModule, IconPipe],
})
export class NavigationBarComponent {
  @Input() layout: 'row' | 'column' = 'row';
  @Input() items?: NavigationBarItem[] | null;
}
