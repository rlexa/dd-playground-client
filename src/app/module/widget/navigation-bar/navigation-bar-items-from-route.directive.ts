import {ChangeDetectorRef, Directive, OnDestroy, OnInit, inject} from '@angular/core';
import {ActivatedRoute, Data} from '@angular/router';
import {DoneSubject, RxCleanup} from 'dd-rxjs';
import {filter, map, takeUntil} from 'rxjs/operators';
import {NavigationBarComponent, NavigationBarItem} from './navigation-bar.component';

export interface NavigationBarItemsData extends Data {
  navigationBarItems: NavigationBarItem[];
}

@Directive({selector: '[appNavigationBarItemsFromRoute]', standalone: true})
export class NavigationBarItemsFromRouteDirective implements OnDestroy, OnInit {
  private readonly activatedRoute = inject(ActivatedRoute);
  private readonly navigationBarComponent = inject(NavigationBarComponent);
  private readonly changeDetectorRef = inject(ChangeDetectorRef);

  @RxCleanup() private readonly done$ = new DoneSubject();

  destroy() {}
  ngOnDestroy() {
    this.destroy();
  }

  ngOnInit() {
    this.activatedRoute.data
      .pipe(
        filter(() => !!this.navigationBarComponent),
        map((data: NavigationBarItemsData) => data?.navigationBarItems ?? null),
        takeUntil(this.done$),
      )
      .subscribe((items) => {
        this.navigationBarComponent.items = items;
        this.changeDetectorRef.markForCheck();
      });
  }
}
