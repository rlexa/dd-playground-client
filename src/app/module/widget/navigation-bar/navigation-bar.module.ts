import {CommonModule} from '@angular/common';
import {NgModule} from '@angular/core';
import {MatButtonModule} from '@angular/material/button';
import {MatToolbarModule} from '@angular/material/toolbar';
import {MatTooltipModule} from '@angular/material/tooltip';
import {RouterModule} from '@angular/router';
import {IconPipeModule} from '../../pipe/icon';
import {NavigationBarComponent, NavigationBarItem} from './navigation-bar.component';

@NgModule({
  imports: [CommonModule, RouterModule, MatButtonModule, MatToolbarModule, MatTooltipModule, IconPipeModule],
  exports: [NavigationBarComponent],
  declarations: [NavigationBarComponent],
})
class NavigationBarModule {}

export {NavigationBarModule, NavigationBarComponent, NavigationBarItem};
