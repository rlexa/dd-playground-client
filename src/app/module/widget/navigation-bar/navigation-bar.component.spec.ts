import {ComponentFixture, TestBed} from '@angular/core/testing';
import {MatButton} from '@angular/material/button';
import {MatToolbar} from '@angular/material/toolbar';
import {MatTooltip} from '@angular/material/tooltip';
import {RouterTestingModule} from '@angular/router/testing';
import {MockComponents, MockDirectives, MockPipe} from 'ng-mocks';
import {detectChanges, overrideForChangeDetection} from 'src/app/test';
import {IconPipe} from '../../pipe/icon';
import {NavigationBarComponent, NavigationBarItem} from './navigation-bar.component';

describe('NavigationBarComponent', () => {
  let fixture: ComponentFixture<NavigationBarComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule],
      declarations: [
        NavigationBarComponent,
        MockComponents(MatToolbar, MatButton),
        MockDirectives(MatTooltip),
        MockPipe(IconPipe, (val) => `icon ${val}`),
      ],
      providers: [],
    })
      .overrideComponent(NavigationBarComponent, overrideForChangeDetection)
      .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(NavigationBarComponent);
    detectChanges(fixture);
  });

  it('creates instance', () => expect(fixture.componentInstance).toBeTruthy());
  it(`renders`, () => expect(fixture).toMatchSnapshot());

  describe(`with items`, () => {
    beforeEach(() => {
      fixture.componentInstance.items = [1, 2, 3].map<NavigationBarItem>((ii) => ({
        icon: `icon${ii}`,
        label: `label${ii}`,
        route: `route${ii}`,
      }));
      detectChanges(fixture);
    });

    it(`renders`, () => expect(fixture).toMatchSnapshot());

    describe(`with layout column`, () => {
      beforeEach(() => {
        fixture.componentInstance.layout = 'column';
        detectChanges(fixture);
      });

      it(`renders`, () => expect(fixture).toMatchSnapshot());
    });
  });
});
