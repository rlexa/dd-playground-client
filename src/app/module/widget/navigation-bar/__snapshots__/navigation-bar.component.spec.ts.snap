// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBarComponent renders 1`] = `
<app-navigation-bar
  items="null"
  trackByIndex={[Function Function]}
>
  <mat-toolbar
    color="primary"
  >
    <span
      class="mat-toolbar-filler"
    />
    
  </mat-toolbar>
</app-navigation-bar>
`;

exports[`NavigationBarComponent with items renders 1`] = `
<app-navigation-bar
  items={[Function Array]}
  trackByIndex={[Function Function]}
>
  <mat-toolbar
    color="primary"
  >
    <span
      class="mat-toolbar-filler"
    />
    
    
    <nav
      class="mat-nav-list"
    >
      
      <a
        href="/route1"
        mat-button=""
        routerlinkactive="active"
        style="min-width: auto;"
      >
        <i
          class="material-icons transition-transform active-scale"
        >
          icon icon1
        </i>
      </a>
      
      <a
        href="/route2"
        mat-button=""
        routerlinkactive="active"
        style="min-width: auto;"
      >
        <i
          class="material-icons transition-transform active-scale"
        >
          icon icon2
        </i>
      </a>
      
      <a
        href="/route3"
        mat-button=""
        routerlinkactive="active"
        style="min-width: auto;"
      >
        <i
          class="material-icons transition-transform active-scale"
        >
          icon icon3
        </i>
      </a>
      
    </nav>
    <span
      class="mat-toolbar-filler"
    />
  </mat-toolbar>
</app-navigation-bar>
`;
